### ISSUE TITLE
config value escaping

### ISSUE THREAD

**Issue by mhahn** [deis/deis#2980]
I had something like the following config set for DATABASE_URL:

```
$ deis config
=== services Config
DATABASE_URL                postgres://user:EbMCMW0^`ToWS7-v@example.us-east-1.rds.amazonaws.com:5432
```

when I was trying to push I got this error:

```
$ git push deis master
Counting objects: 4, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 411 bytes | 0 bytes/s, done.
Total 4 (delta 1), reused 0 (delta 0)
/home/git/builder: eval: line 117: unexpected EOF while looking for matching ``'
/home/git/builder: eval: line 118: syntax error: unexpected end of file
To ssh://git@deis.example.net:2222/services.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to 'ssh://git@deis.example.net:2222/services.git'
```

removing the "`" character fixed the issue.

**Comment by bacongobbler** [comment]
That indicates an error [here](https://github.com/deis/deis/blob/master/builder/image/templates/builder#L117) which most likely means there's some parsing error in [`parse-string`](https://github.com/deis/deis/blob/master/builder/image/templates/builder#L30).

**Comment by carmstrong** [comment]
Probably the tick
```
`
```
 is being interpolated by the shell.

**Comment by aledbf** [comment]
@bacongobbler I think that error is here [builder](https://github.com/deis/deis/blob/master/builder/image/templates/builder#L92) in the echo of `$RESPONSE`

**Comment by bacongobbler** [comment]
I can reliably reproduce with the following:

```
$ eval $(parse-string 'docker run -e DATABASE_URL=postgres://user:foobar`a@example.us-east-1.rds.amazonaws.com:5432')
```

So we need to properly escape the backtick before executing the process.

**Comment by technosophos** [comment]
Might we be better served by using base64 encoding to pass these values around, and simply adding a decoder at the last mile?

**Comment by bacongobbler** [comment]
That would certainly help, but not necessarily bacwards-compatible unless we do a data migration (which is completely possible)

**Comment by bacongobbler** [comment]
Sorry, poor English. Yes, that would help

**Comment by technosophos** [comment]
For v2, I'd like us to switch this feature to using Kubernetes secrets for envvar storage. That will automatically handle all of the escaping during storage, etc.

**Comment by gabrtv** [comment]
@technosophos :100: though we will need to expose file-based secrets as envvars for application compatibility. Thankfully there some good work being done upstream to help with this: https://github.com/kubernetes/kubernetes/blob/master/docs/design/secrets.md#deferral-consuming-secrets-as-environment-variables

**Comment by technosophos** [comment]
Yup, for sure, buildpack-based ones must have them as environment variables. I guess we should decide whether that's the best course for Dockerfile builds. There's no technical reason why we couldn't do it across the board, though.

**Comment by bacongobbler** [comment]
I believe this is resolved now in v2 at the very least, but we haven't released a stable version of that yet.

**Comment by bacongobbler** [comment]
Actually this should also might be resolved in v1.12

**Comment by mboersma** [comment]
This is fixed by #5006, in the upcoming Deis v1.13.0 release.
